Auto upgrade feature

DishantK1807 opened the pull request:
Issue : #162 

![update01](https://cloud.githubusercontent.com/assets/7828295/17595454/c4ff6704-600a-11e6-8ca5-91191f860be4.PNG)
Enable disable update in the UI 

![update02](https://cloud.githubusercontent.com/assets/7828295/17595455/c5002d06-600a-11e6-8a3a-3fec9584a15d.PNG)
System updated to latest version 

![update03](https://cloud.githubusercontent.com/assets/7828295/17595453/c4d6db86-600a-11e6-86dc-d4069a2c24d7.PNG)
Check for updates enabled

DishantK1807 committed:
Created Update script

DishantK1807 committed:
Created version file

DishantK1807 committed:
Update update.sql

DishantK1807 committed:
Added update feature

DishantK1807 committed:
Updated setting_model for auto_update

DishantK1807 committed:
Implemented automatic check for updates

DishantK1807 committed:
changed the messages displayed

DishantK1807 committed:
Updated documentation for AutoUpdates

hongquan commented on includes/controller/user_settings_controller.php:
```diff
@@ -193,6 +194,41 @@ function user_settings() {
+      $ok = false;
+      $msg .= error(_("The system is already Up-to date with the version on GitHub."), true);
+    }
+    if($ok) {
```
Coding style error

DishantK1807 replied:
I'll add a blank line before the `if($ok)`

DishantK1807 replied:
I guess only that's the error, or did I missed something else?

hongquan replied:
https://www.drupal.org/coding-standards#controlstruct

hongquan commented on includes/controller/user_settings_controller.php:
```diff
@@ -193,6 +194,41 @@ function user_settings() {
+    $online_ver = file_get_contents("https://raw.githubusercontent.com/DishantK1807/engelsystem/auto-update/Version.txt");
+    $current_ver = file_get_contents(" ./Version.txt");
+    if (strcmp($current_ver, $online_ver) != 0) {
+      echo shell_exec("sudo sh ./update.sh");
```
How is sudo password passed to command?

DishantK1807 replied:
it took in the terminal, the second time the script was run. Although it is not required. I'll remove it.

hongquan commented on update.sh:
```diff
@@ -0,0 +1,13 @@
+git remote add upstream http://www.github.com/fossasia/engelsystem.git
+
+# Pull the updates
+git pull upstream development
```
You check for version number in your repo: https://raw.githubusercontent.com/DishantK1807/engelsystem/auto-update/Version.txt
but you grab the new code from _fossasia/engelsystem_ repo?

DishantK1807 replied:
I was to change the link after merging the code, after that file is merged to the branch.

hongquan replied:
Then how do you test? Not matter what number you put to Version.txt, the `git pull upstream development` command still don't get the new code.

DishantK1807 replied:
I test on my local system, pulling the changes from my repository making the changes in the Version.txt and running the script file. [which includes pulling from my repository]

hongquan replied:
> which includes pulling from my repository

Pulling from your repository? This code

```
git remote add upstream http://www.github.com/fossasia/engelsystem.git
git pull upstream development
```

shows that it will pull from http://www.github.com/fossasia/engelsystem.git, not your repo.

DishantK1807 replied:
<pre>
#!/usr/bin/env bash

# Authors:
#   Dishant Khanna <dishant.khanna1807@gmail.com> [FOSSASIA]

# Description:
#   An update bash script for engelsystem

# Add upstream 
git remote add upstream http://www.github.com/DishantK1807/engelsystem.git

# Pull the updates
git pull upstream auto-update
</pre>


This is the script that I used for testing on my local server. It includes pulling from my repository, pull the auto-update branch. I changed the repo in the script while pushing the code to fossasia.

hongquan replied:
What I complaint here is the "inconsistency". You check for new version in one repo, and download new code from another repo!

If your work is not ready to merge (when you still use the links to your repo), don't create pull request.

And don't say that

> change the link after merging the code

The code to be merged, should be the stable one. You cannot make change after it is merged. If you found the error before it is merged, fix it right now, before it is merged (this rule is not applied for the error which is discovered after merging).

DishantK1807 replied:
Sorry. I'll change the script to my repo in this PR. And make another PR after it is merged for changing the script to pull from fossasia repo.
Thank you

hongquan replied:
If you know that the URL will be changed, change it before merging. New PR is for mistake that you find out **after merge**.

I repeat again, the code to be merged in PR should be the "stable" one.

DishantK1807 replied:
But the problem is there is no Version.txt file in the Fossasia repo.

hongquan replied:
Isn't "Version.txt" included in this PR as well? After PR is merged, the Version.txt will exist (If not, this PR won't be merged).

DishantK1807 committed:
Fix: Coding Style errors

DishantK1807 committed:
Changed links

hongquan commented on update.sh:
```diff
@@ -6,8 +6,8 @@
 
-# Add upstream 
+# Add upstream
 git remote add upstream http://www.github.com/fossasia/engelsystem.git
```
Do you notice that "origin" is also "http://www.github.com/fossasia/engelsystem.git"?
No need to define a "remote" with the same URL as "origin".

DishantK1807 replied:
If a person Forks the project and makes some minor changes to it and uses it on his/her system, then remote would be required. 
In the case of forked repositories, "remote" will be different from "origin"

hongquan replied:
If he forks the project and make somes changes, he should update from his repo, not FOSSASIA repo, because who can guarantee that new change in FOSSASIA won't break his system? The code update in FOSSASIA repo is only tested against the code in FOSSASIA repo, not against code of someone else. And who guarantee that new update from FOSSASIA won't have conflict with the change he made in his repo?

DishantK1807 replied:
I'll remove "remote" and pull from "origin"
Thank you.

DishantK1807 committed:
change upstream to origin

hongquan commented:
Why do you still let this happen:

![selection_136](https://cloud.githubusercontent.com/assets/314607/17799506/657e9f4c-6605-11e6-8006-8c72d2225e9a.png)

when I told you to config your code editor to fix it automatically?

(You don't need to make a commit to fix it right now. Don't repeat this error in the future)

DishantK1807 committed:
resolving conflicts

DishantK1807 commented:
@mariobehling : Can we merge this PR?

hongquan commented:
How do you deal with the upgrade which includes change in database schema? Just update the code will break the system!

DishantK1807 commented:
@hongquan: I can make and `upgrade.sql` file which will contain the changes in the database scheme. In `upgrade.sql`  I'll implement checks to see if a table already exists and make database changes accordingly.

hongquan commented:
> I can make and upgrade.sql file which will contain the changes in the database scheme

Where in this PR that you run the _upgrade.sql_ file, if that file exists?

Show me the screenshot that you have tested this feature in your localhost (with and without database change)

Can you explain this line:

```php
echo shell_exec("sudo sh ./update.sh");
```

Why `echo`? Are you aware that in the "controller" PHP scripts, the content to return to browser is always wrapped in the functions like `div`, `page_link_to` and never returned by `echo`?

Why `shell_exec` comes with `sh`? Quoted from http://php.net/manual/en/function.shell-exec.php:

> shell_exec — Execute command via shell...

The path `./update.sh` won't work with new URL of engelsystem (changed from `http://example.com/engelsystem/public/` to `http://example.com/ `). Talk with @kamishettysreeja25 how she implemented the change and hope you will understand why that path will not work.

What if `sudo` ask for password? How user pass the password in?

DishantK1807 commented:
>The path ./update.sh won't work with new URL of engelsystem (changed from http://example.com/engelsystem/public/ to http://example.com/). Talk with @kamishettysreeja25 how she implemented the change and hope you will understand why that path will not work.

when she already implemented this in the install process, I don't think there would be any need to change the URL of the Engelsystem again.

hongquan commented:
> I found this on stackoverflow.com

Don't just copy and paste from stackoverflow. You have to understand what you are doing. This is not the first time I remind you to read documentation and understand the command you are using.

sreejakshetty commented on includes/controller/user_settings_controller.php:
```diff
@@ -187,12 +188,48 @@ function user_settings() {
         $ok = false;
-      if($ok){
+
+      if ($ok){
```
Indentation error: It should be `if ($ok) {`

hongquan replied:
Good. Thanks!

sreejakshetty commented on includes/controller/user_settings_controller.php:
```diff
@@ -187,12 +188,48 @@ function user_settings() {
+    $ok = true;
+
+    $online_ver = file_get_contents("https://raw.githubusercontent.com/fossasia/engelsystem/master/Version.txt");
+    $current_ver = file_get_contents(" ./Version.txt");
```
All the links are wrong. According to the work flow we include all the files in public/index.php. We need to give path relative to public folder.
Please refer this code https://github.com/fossasia/engelsystem/blob/development/includes/model/import_model.php.

sreejakshetty commented on includes/controller/user_settings_controller.php:
```diff
@@ -187,12 +188,48 @@ function user_settings() {
+    $online_ver = file_get_contents("https://raw.githubusercontent.com/fossasia/engelsystem/master/Version.txt");
+    $current_ver = file_get_contents(" ./Version.txt");
+    if (strcmp($current_ver, $online_ver) != 0) {
+      echo shell_exec("sudo sh ./update.sh");
```
This also don't work.

sreejakshetty replied:
Instead of echo you can use exec command
http://php.net/manual/en/function.exec.php

DishantK1807 commented:
No
removing `sudo` won't cause any problems. I just added it for the system which won't perform any operations without superuser permissions. 
the script is running fine even without sudo.

DishantK1807 committed:
changed links

DishantK1807 committed:
Fix: upgrade procedure

DishantK1807 commented on includes/controller/user_settings_controller.php:
```diff
@@ -199,9 +199,14 @@ function user_settings() {
+      if(strcmp($current_ver, "Version: 1.0") == 0){
+        $upgrade_table = '../db/upgrade_01.sql';
+        upgrade_tables($upgrade_table);
+      }
```
Similarly we can create `upgrade_XX.sql` for every release and compare the version numbers of the of the local system.

I'll add in the documentation too, how to upgrade the database(with a sample code explaining everything) so that users who'll be working on it in future can integrate the changes in the database.

Eg. If the current version in the local system is `3.0` and on GitHub it is `5.0` then in the if statement both `Upgrade_04.sql` and `upgrade_05.sql` will run.
Will update the documentation for the same

hongquan commented on public/index.php:
```diff
@@ -28,6 +28,16 @@
+  $check_autoupdate_enable = check_AutoUpdate();
+
+  if ($check_autoupdate_enable == true) {
+  $online_ver = file_get_contents("https://raw.githubusercontent.com/fossasia/engelsystem/master/Version.txt");
```
Indentation error!

hongquan commented:
Now the _update.sh_ script contains only 1 command. Let's get rid of that file.

By creating _upgrade_XX.sql_, you are repeating the work done by the "database migration" feature. Please look for a way to reuse that feature.

DishantK1807 commented:
>By creating upgrade_XX.sql, you are repeating the work done by the "database migration" feature. Please look for a way to reuse that feature.

@hongquan: can we implement that as a Fix/Improvement and merge this PR? I am looking for a way to reuse that feature.

>Now the update.sh script contains only 1 command. Let's get rid of that file.

I suggest not to get rid of that file because we will require that while running the migrations.

DishantK1807 committed:
Fix: Indentataion errors

hongquan commented:
> I suggest not to get rid of that file because we will require that while running the migrations

What is required is the commands in that file. Because it has only 1 command, you should pass that command to your `shell_exec`, instead of creating 1 more script for just that command and pass the script to `shell_exec`.

> can we implement that as a Fix/Improvement and merge this PR

The PR is merged after it is tested to show that it is stable and not break the system. You still don't provide the screenshots to show that you have tested with update + database. It is not proved to be stable to merge yet!

DishantK1807 committed:
resolving conflicts

DishantK1807 committed:
Removed the Shell file

DishantK1807 commented:
The Apache log:
https://cryptbin.com/hUAN#369ff3ad86f47434ca748077dfc6773e

Contents of `upgrade_01.sql`:
<pre> INSERT INTO `Privileges` (`id`, `name`, `desc`) VALUES (45, 'install23', 'Install Engelsystem');
</pre>

Tested the system for a random database upgrade.

![upgraded](https://cloud.githubusercontent.com/assets/7828295/17879300/94624fc8-690e-11e6-9926-6f3d1d40b14d.png)
System Upgraded successfully

DishantK1807 commented:
Will update the documentation.
Thank you.

hongquan commented:
Read my https://github.com/fossasia/engelsystem/pull/220#issuecomment-241208074 again:

> upgrade which includes change in database schema

Change in **database schema**. Inserting data to database is not a change in schema.

hongquan commented:
When you get the Apache error log, did you look into it to discover error and try to fix?

DishantK1807 commented:
> When you get the Apache error log, did you look into it to discover error and try to fix?

I will be working on fixing those errors, most of them are not because of this feature.

> Change in database schema. Inserting data to database is not a change in schema.

I'll test for change in schema as well

DishantK1807 commented:
The Apache log:
http://cryptb.in/NEowIW#ad4eed3bf192e327e1c31fed2df8e234

content of `upgrade_01.sql`
<pre>ALTER TABLE `Room`
  ADD `e_id` int(11) NOT NULL; </pre>

![upgraded](https://cloud.githubusercontent.com/assets/7828295/17880020/17030940-6914-11e6-82b3-a6e3a4ee8e35.png)

System Upgraded successfully.

hongquan commented:
I reminded you to look into Apache error log, but you still ignore this:

> [Tue Aug 23 09:26:00.998777 2016] [:error] [pid 3074] [client 127.0.0.1:41481] PHP Warning:  file_get_contents(raw.githubusercontent.com/fossasia/engelsystem/master/Version.txt): failed to open stream: HTTP request failed! HTTP/1.1 404 Not Found\r\n in /var/www/html/engelsystem/public/index.php on line 34
[Tue Aug 23 09:26:00.998832 2016] [:error] [pid 3074] [client 127.0.0.1:41481] PHP Warning:  file_get_contents( ../Version.txt): failed to open stream: No such file or directory in /var/www/html/engelsystem/public/index.php on line 35

Look into kamishettysreeja25 PR about "Wordpress-like installation" to learn how to prevent the other errors. And also you can merge "development" branch to your branch to update with the fix she has done.

hongquan commented:
```php
      if(strcmp($current_ver, "Version: 1.0") == 0){
        $upgrade_table = '../db/upgrade_01.sql';
        upgrade_tables($upgrade_table);
      }
```

If _Version.txt_ contains higher number, like `Version: 2.0`, the database migration won't happen, right? And then the system will break?

DishantK1807 commented:
> If Version.txt contains higher number, like Version: 2.0, the database migration won't happen, right? And then the system will break?

I explained that in the Line comment, https://github.com/fossasia/engelsystem/pull/220/files#r75583290
That we would need to update code the similar way.

I'll add that to the documentation/wiki

DishantK1807 commented:
Apache log:
http://cryptb.in/C7hr#10db833189ee8869546c86359431e522

Fixed the mentioned error. the error was due to the wrong link in index.php
Changed the link in local repo and tested.

![upgraded](https://cloud.githubusercontent.com/assets/7828295/17880470/d7f25c10-6918-11e6-806a-20c8ece6ce1d.png)

hongquan commented:
What? You ask user to modify source code? He/she has to find which version his/her website is, and which version the new update is, and has to implement the whole `if` code block? Then how come this feature is called "auto upgrade"? This feature has been existed in many PHP web system like Wordpress, Drupal, Piwik but I have never seen one asking user to modify source code to upgrade.

Needless to say that user doesn't know when to change, because:

- If user modify before the upgrade process happen, the "git pull" command will stop if the new update contains update for the file user has modified. Just find out the version number is already headache for user.
- If user modify after "git pull", no chance. The process won't pause for user to intervene.

DishantK1807 commented:
> What? You ask user to modify source code? He/she has to find which version his/her website is, and which version the new update is, and has to implement the whole if code block? Then how come this feature is called "auto upgrade"? This feature has been existed in many PHP web system like Wordpress, Drupal, Piwik but I have never seen one asking user to modify source code to upgrade.
Needless to say that user doesn't know when to change, because:
If user modify before the upgrade process happen, the "git pull" command will stop if the new update contains update for the file user has modified. Just find out the version number is already headache for user.
If user modify after "git pull", no chance. The process won't pause for user to intervene.

No. not the user. User won't work on the code.
The person contributing the code. 
Will add this to the wiki as a condition when the PR is merged.

hongquan commented:
> The person contributing the code. 
> If the current version in the local system is 3.0 and on GitHub it is 5.0 then in the if statement both Upgrade_04.sql and upgrade_05.sql will run.
Will update the documentation for the same

Then how the developer know which version the local system is? Assume you are releasing v5, but Mario has a website with v4, and sister Phuc has a website with v1. How the developer implement that `if` block to satisfy both?
